### PR TITLE
Process pending donation with PayPal  webhook

### LINKED ISSUE #5062: Pending donations should be resolved via webhooks
@JasonTheAdams commented on [Tue Aug 11 2020](https://github.com/impress-org/gateway-oauth/issues/11)

## User Story

As a donor with a pending donation, I want to see when my donation has been approved or denied, so that I know how it finished.

## Details

Payments in PayPal sometimes take a bit longer to be approved. In these cases the Payment goes into a pending state. Since this can take some time, it's not a synchronous process. In our case, the donation is completed but shows as pending. This is currently in place.

What's not in place, is that the pending state is ultimately resolved by webhook, being marked as completed or denied. We need to support the [following webhooks](https://developer.paypal.com/docs/api-basics/notifications/webhooks/event-names/#v2) and update the donation accordingly:
- PAYMENT.CAPTURE.COMPLETED
- PAYMENT.CAPTURE.DENIED

## Acceptance Criteria

-   [x] Pending payments can be marked as completed via the PAYMENT.CAPTURE.COMPLETED webhook
-   [x] Pending payments can be marked as denied via the PAYMENT.CAPTURE.DENIED webhook
-   [x] Once a donation is completed or denied, the donor and admin are notified

### PR DESCRIPTION
**Pull request by ravinderk**
Resolves #5062

## Description
This pr handle PAYMENT.CAPTURE.COMPLETED and PAYMENT.CAPTURE.DENIED webhook of PayPal.

## Affects
Existing payment refund webhook group with other payment webhook in the `PayPalCommerce` directory. Update `PaymentCaptureRefunded:processEvent` function to check donation status also if donation status already set to `refund` then exit process.

## Visuals
Note: do not confuse with `completed` and `denied` note in same donation. These notes generated because I simulate webhook on the same donation.

image

## Pre-review Checklist
-   [x] Acceptance criteria satisfied and marked in related issue
-   [ ] Tests included
-   [ ] Keyboard accessible
-   [ ] Screen reader accessible
-   [x] Relevant `@since` tags included in DocBlocks
-   [ ] Changelog updated
-   [ ] Labels applied if docs are needed
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

## Testing Instructions
Grab webhooks sample data from PayPal webhook simulator and then trigger webhook with postman: https://developer.paypal.com/developer/webhooksSimulator/ 

Webhook sample:
```
{
  "id": "WH-4SW78779LY2325805-07E03580SX1414828",
  "create_time": "2019-02-14T22:20:08.370Z",
  "resource_type": "capture",
  "event_type": "PAYMENT.CAPTURE.DENIED",
  "summary": "A AUD 2.51 AUD capture payment was denied",
  "resource": {
    "amount": {
      "currency_code": "AUD",
      "value": "2.51"
    },
    "seller_protection": {
      "status": "ELIGIBLE",
      "dispute_categories": [
        "ITEM_NOT_RECEIVED",
        "UNAUTHORIZED_TRANSACTION"
      ]
    },
    "update_time": "2019-02-14T22:20:01Z",
    "create_time": "2019-02-14T22:18:14Z",
    "final_capture": true,
    "seller_receivable_breakdown": {
      "gross_amount": {
        "currency_code": "AUD",
        "value": "2.51"
      },
      "net_amount": {
        "currency_code": "AUD",
        "value": "2.51"
      }
    },
    "links": [
      {
        "href": "https://api.paypal.com/v2/payments/captures/0LA63881BJ6253222",
        "rel": "self",
        "method": "GET"
      },
      {
        "href": "https://api.paypal.com/v2/payments/captures/0LA63881BJ6253222/refund",
        "rel": "refund",
        "method": "POST"
      },
      {
        "href": "https://api.paypal.com/v2/payments/authorizations/0LA63881BJ6253222",
        "rel": "up",
        "method": "GET"
      }
    ],
    "id": "0LA63881BJ6253222",
    "status": "DECLINED"
  },
  "links": [
    {
      "href": "https://api.paypal.com/v1/notifications/webhooks-events/WH-4SW78779LY2325805-07E03580SX1414828",
      "rel": "self",
      "method": "GET",
      "encType": "application/json"
    },
    {
      "href": "https://api.paypal.com/v1/notifications/webhooks-events/WH-4SW78779LY2325805-07E03580SX1414828/resend",
      "rel": "resend",
      "method": "POST",
      "encType": "application/json"
    }
  ],
  "event_version": "1.0",
  "resource_version": "2.0"
}
```

Make sure you disable event signature validation from code before testing webhooks:
image

### CODE REVIEW

File: src/PaymentGateways/PayPalCommerce/Webhooks/Listeners/PayPalCommerce/PaymentCaptureCompleted.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ * @package Give\PaymentGateways\PayPalCommerce\Webhooks\Listeners\PayPalCommerce
+ *
+ * @since 2.8.0
+ */
+class PaymentCaptureCompleted extends PaymentEventListener {
+	/**
+	 * @inheritDoc
+	 */
+	public function processEvent( $event ) {
+		$paymentId = $this->getPaymentFromRefund( $event->resource, 'self' );
```
Comment (JasonTheAdams) [inline src/PaymentGateways/PayPalCommerce/Webhooks/Listeners/PayPalCommerce/PaymentCaptureCompleted.php:16]: @ravinderk I'm really confused by this. Why do you need to get the Payment from the API... again? By using the "self" link you're literally just grabbing the Payment _again_. In this context, `$event->resource` **is** a Payment, so all you need is `$event->resource->id`.

Comment (ravinderk) [reply src/PaymentGateways/PayPalCommerce/Webhooks/Listeners/PayPalCommerce/PaymentCaptureCompleted.php]: fixed

**Comment by JasonTheAdams** [comment]
@ravinderk Is there a credit card number or something that will put a payment into a pending state and then complete or deny it? I'd like to test this using real data, not just the simulator.

**Comment by ravinderk** [comment]
I test this by simulating webhook with postmen but as per suggestion, it is best to test real world scenarios. We are waiting on the PayPal team to guide us on `how can we process the payment with PayPal which auto denied`.

slack discussion: https://givewp.slack.com/archives/C015XD31J5S/p1597338771041600

**Pull request merged**